File: FactoryMethod/src/com/ash/maze/Door.java
package com.ash.maze;

public class Door implements MapSite {
    private final Room firstRoom;
    private final Room secondRoom;
    private boolean isOpen;

    public Door(Room firstRoom, Room secondRoom) {
        this.firstRoom = firstRoom;
        this.secondRoom = secondRoom;
        this.isOpen = false;
    }

    void open() {
        this.isOpen = true;
    }

    void close() {
        this.isOpen = false;
    }

    boolean isOpen() {
        return this.isOpen;
    }

    @Override
    public void enter() {

    }

    @Override
    public String toString() {
       return "Regular door between room: " + this.firstRoom.number() + " and room: " + this.secondRoom.number();
    }
}

File: maze/Door.cpp
#include "Door.h"

Door::Door(Room* firstRoom, Room* secondRoom) {
    _firstRoom = firstRoom;
    _secondRoom = secondRoom;
    _isOpen = false;
}

void Door::Enter() {
    cout << "Door is open FROM " << endl; 
    _isOpen = !_isOpen;
}

bool Door::isOpen() {
    return _isOpen;
}

string Door::desc() {
    return "door";
}
